### PR TITLE
Add Recipe presenting real world use case

### PR DESCRIPTION
**Pull request by Wosin**
I was thinking about something that presents a concept in a simple manner but also captures the essence of update based on the effectful result, but not too sure if this example is the best.

### CODE REVIEW

File: docs/recipes.md [hunk 1]
```diff
@@ -90,3 +90,61 @@ object Server extends IOApp.Simple {
...
+However, if we want to update our ref using result of an effect `Ref` will usually not be enough and we need a more powerful construct to achieve that. 
+In cases like that we can use the [AtomicCell](std/atomic-cell.md) that can be viewed as a synchronized `Ref`.
+
+The most typical example for `Ref` is the concurrent counter, but what if the update function for our counter would be effectful?
+
+Assume we have the following function:
+
+```scala mdoc:silent
+def update(input: Int): IO[Int] =
+  IO.defer(input + 1 )
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
  IO(input + 1 )
```

:-) I think you meant `delay`, but `apply` is even more concise.

File: docs/recipes.md [hunk 2]
```diff
@@ -90,3 +90,61 @@ object Server extends IOApp.Simple {
...
+```
+If we want to concurrently update a variable using our `update` function it's not possible to do it directly with `Ref`, luckily `AtomicCell` has `evalUpdate`:
+
+```scala mdoc:silent
+class Server(atomicCell: AtomicCell[IO, Int]) {
+  def update(input: Int): IO[Int] =
+    IO.defer(input + 1 )
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
    IO(input + 1 )
```

File: docs/recipes.md [hunk 3]
```diff
@@ -90,3 +90,61 @@ object Server extends IOApp.Simple {
...
+
+  def performUpdate(input: Int): IO[Int] = 
+    atomicCell.evalUpdate(i => update(i))
+}
+```
+So, let's assume that we want to implement very simple rate limiting proxy, that will only allow some specific number of requests and then return failure if this limit is surpassed.
+
+The code could look like below:
+```scala mdoc:silet
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
```scala mdoc:silent
```

File: docs/recipes.md [hunk 4]
```diff
@@ -90,3 +90,61 @@ object Server extends IOApp.Simple {
...
+}
+
+object StubService extends Service {
+  def query(request: Request): IO[Response] = IO.pure(Response("Request Succeeded."))
+}
+
+class RateLimiterProxy(limit: Int, cell: AtomicCell[IO, Int], service: Service){
+  def limit[A](request: IO[Request]): IO[Response] = {
+    cell.evalModify{ requests =>
+      if(requests >= limit)
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
      if (requests >= limit)
```

File: docs/recipes.md [hunk 5]
```diff
@@ -90,3 +90,61 @@ object Server extends IOApp.Simple {
...
+  def query(request: Request): IO[Response]
+}
+
+object StubService extends Service {
+  def query(request: Request): IO[Response] = IO.pure(Response("Request Succeeded."))
+}
+
+class RateLimiterProxy(limit: Int, cell: AtomicCell[IO, Int], service: Service){
+  def limit[A](request: IO[Request]): IO[Response] = {
+    cell.evalModify{ requests =>
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
    cell evalModify { requests =>
```

**Review by djspiewak — CHANGES_REQUESTED** [review]
Should we discuss the `modify` + `flatten` idiom as well?

**Comment by Wosin** [comment]
Hey, I've actually found simpler example, let me know if that works, I think RateLimiter was a little farfetched.

File: docs/recipes.md [hunk 6]
```diff
@@ -90,3 +90,67 @@ object Server extends IOApp.Simple {
...
+  def performUpdate(input: Int): IO[Int] = 
+    atomicCell.evalUpdate(i => update(i))
+}
+```
+
+To better present real-life use-case scenario, let's imagine that we have a `Service` that performs some HTTP request to external service holding exchange rates over time:
+
+```scala mdoc:silent
+case class Response(exchangeRate: Double)
+trait Service{
```
Comment (iRevive) [inline docs/recipes.md]: To keep things consistent. 

```suggestion
trait Service {
```

File: docs/recipes.md [hunk 7]
```diff
@@ -90,3 +90,67 @@ object Server extends IOApp.Simple {
...
+  def query(): IO[Response]
+}
+
+object StubService extends Service{
```
Comment (iRevive) [inline docs/recipes.md]:
```suggestion
object StubService extends Service {
```

File: docs/recipes.md [hunk 8]
```diff
@@ -90,3 +90,67 @@ object Server extends IOApp.Simple {
...
+```
+To simplify we model the `StubService` to just return some random `Double` values.
+
+Now, say that we want to have a cache that holds the highest exchange rate that is ever returned by our service, we can have the proxy implementation based on `AtomicCell` like below:
+
+```scala mcdoc:silent
+class MaxProxy(atomicCell: AtomicCell[IO, Double], requestService: Service) {
+
+  def queryCache(): IO[Response] = {
+    atomicCell evalModify (current => {
```
Comment (iRevive) [inline docs/recipes.md]:
```suggestion
    atomicCell evalModify { current => 
```

File: docs/recipes.md [hunk 9]
```diff
@@ -90,3 +90,67 @@ object Server extends IOApp.Simple {
...
+      requestService.query().map(result =>
```
Comment (iRevive) [inline docs/recipes.md]: I do not have a strong opinion there. A curly bracket feels more idiomatic for a multiline block, but both options are valid.

```suggestion
      requestService.query().map { result => 
```

File: docs/recipes.md [hunk 10]
```diff
@@ -90,3 +90,75 @@ object Server extends IOApp.Simple {
 
 In this example, `longRunningTask` is started in the background.
 The server returns to the client without waiting for the task to finish.
+
+## Atomically update a Ref with result of an effect
+
+Cats effect provides [Ref](std/ref.md), that we can use to model mutable concurrent reference.
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
Cats Effect provides [Ref](std/ref.md), that we can use to model mutable concurrent reference. 
```

File: docs/recipes.md [hunk 11]
```diff
@@ -90,3 +90,75 @@ object Server extends IOApp.Simple {
...
+However, if we want to update our ref using result of an effect `Ref` will usually not be enough and we need a more powerful construct to achieve that. 
+In cases like that we can use the [AtomicCell](std/atomic-cell.md) that can be viewed as a synchronized `Ref`.
+
+The most typical example for `Ref` is the concurrent counter, but what if the update function for our counter would be effectful?
+
+Assume we have the following function:
+
+```scala mdoc:silent
+def update(input: Int): IO[Int] =
+  IO(input + 1 )
```
Comment (djspiewak) [inline docs/recipes.md]:
```suggestion
  IO(input + 1)
```

File: docs/recipes.md [hunk 12]
```diff
@@ -90,3 +90,75 @@ object Server extends IOApp.Simple {
...
+To simplify we model the `StubService` to just return some random `Double` values.
+
+Now, say that we want to have a cache that holds the highest exchange rate that is ever returned by our service, we can have the proxy implementation based on `AtomicCell` like below:
+
+```scala mcdoc:silent
+class MaxProxy(atomicCell: AtomicCell[IO, Double], requestService: Service) {
+
+  def queryCache(): IO[ServiceResponse] = {
+    atomicCell evalModify { current =>
+      requestService.query().map { result =>
```
Comment (djspiewak) [inline docs/recipes.md]: Just trying to be consistent on this one.

```suggestion
      requestService.query() map { result =>
```

**Review by djspiewak — CHANGES_REQUESTED** [review]
Looks very good! Just a couple minor cleanups and then ready to go

File: docs/recipes.md [hunk 13]
```diff
@@ -90,3 +90,75 @@ object Server extends IOApp.Simple {
...
+  IO(input + 1)
+```
+If we want to concurrently update a variable using our `update` function it's not possible to do it directly with `Ref`, luckily `AtomicCell` has `evalUpdate`:
+
+```scala mdoc:silent
+import cats.effect.std.AtomicCell
+
+class Server(atomicCell: AtomicCell[IO, Int]) {
+  def update(input: Int): IO[Int] =
+    IO(input + 1 )
```
Comment (iRevive) [inline docs/recipes.md]:
```suggestion
    IO(input + 1)
```

File: docs/recipes.md [hunk 14]
```diff
@@ -90,3 +90,75 @@ object Server extends IOApp.Simple {
...
+    .scalaUtilRandom[IO]
+    .flatMap(random => random.nextDouble)
+    .map(ServiceResponse(_))
+}
+```
+To simplify we model the `StubService` to just return some random `Double` values.
+
+Now, say that we want to have a cache that holds the highest exchange rate that is ever returned by our service, we can have the proxy implementation based on `AtomicCell` like below:
+
+```scala mcdoc:silent
```
Comment (iRevive) [inline docs/recipes.md]: One last change I believe :) 
```suggestion
```scala mdoc:silent
```

**Review by djspiewak — APPROVED** [review]
I like this example quite a bit. Nicely done!

**Pull request merged**